include/permissions.hxx

```diff
@@ -0,0 +1,216 @@
+                                         PermissionTypes permission = Read,
+                                         Regions region = AllRegions) const;
+
+  /// Get the highest permission level with which the given variable
+  /// can be accessed in the given region. The second item
+  /// returned indicates the name of the variable or section from
+  /// which the access rights are derived. If there is no matching
+  /// section then it will be an empty string.
+  std::pair<PermissionTypes, std::string>
+  getHighestPermission(const std::string& variable, Regions region = AllRegions) const;
+
+  /// Get a set of variables and regions for which there is the
+  /// specified level of permission to access. If ``highestOnly`` is
+  /// true then it will only include variables/regions for which this
+  /// is the highest permission.
+  ///
+  ///     Permissions example({"test", {Permissions::Nowhere, Permissions::AllRegions,
+  ///     Permissions::AllRegions, Permissions:Nowhere}});
+  ///     // Print variables which can be read
+  ///     for (const auto [varname, region] :
+  ///     example.getVariablesWithPermission(Permissions::Read, false))
+  ///         std::cout << "Variable name: " << varname << ", Region ID: " << region <<
+  ///         "\n";
+  ///     // Print variables which can only be read (not written)
+  ///     for (const auto [varname, region] :
+  ///     example.getVariablesWithPermission(Permissions::Read, true))
+  ///         std::cout << "Variable name: " << varname << ", Region ID: " << region <<
+  ///         "\n";
+  ///
+  /// The above code would write a line of output from the first
+  /// for-loop, but not the second.
+  std::map<std::string, Regions>
+  getVariablesWithPermission(PermissionTypes permission, bool highestOnly = true) const;
+
+  /// Return a string version of the region names
+  static std::string regionNames(const Regions regions);
+
+private:
+  /// Returns the access rights for the most specific entry in this
+  /// object which matches the variable name. If there are no matching
+  /// entries then the result will indicate no access rights. The
+  /// string indicates the name of the variable from which the access
+  /// rights were derived. It will be empty if there are no matching
+  /// entries.
+  std::pair<std::string, AccessRights> bestMatchRights(const std::string& variable) const;
+
+  /// Return a set of access rights where the lower permissions have
+  /// been updated so that they reflect higher permissions (e.g., read
+  /// permission will be set in all cases where write permission was
+  /// set).
+  static AccessRights applyLowerPermissions(const AccessRights& rights);
+
+  std::map<std::string, AccessRights> variable_permissions;
+};
+
+/// Convenience function to return an object expressing that the
+/// variable should have ReadIfSet permissions on all regions.
+std::pair<std::string, Permissions::AccessRights> readIfSet(std::string varname);
```

Comment:
Might be nicer to make the return value a `struct` with named fields?

Also, probably worth putting the definitions `inline` in the header

Reply:
Done.